llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -9262,12 +9265,28 @@ SDValue TargetLowering::expandAVG(SDNode *N, SelectionDAG &DAG) const {
                        DAG.getShiftAmountConstant(1, VT, dl));
   }
 
+  // For scalars, see if we can efficiently extend/truncate to use add+shift.
+  if (VT.isScalarInteger()) {
+    unsigned BW = VT.getScalarSizeInBits();
+    EVT ExtVT = VT.getIntegerVT(*DAG.getContext(), 2 * BW);
+    if (isTypeLegal(ExtVT) &&
+        ((!IsSigned && isZExtFree(VT, ExtVT)) || isTruncateFree(ExtVT, VT))) {
```

Comment:
I guess there is no `isSExtFree` that you can use? You just assume the sext is free if the truncate is free?

Reply:
TBH the isZExtFree is useless as well as isTruncateFree works fine for the x86 cases